etna/packages/etna-py/src/mountetna/metis.py

```diff
@@ -466,10 +466,25 @@ def upload_file(
             self.create_folder(project_name, bucket_name, os.path.dirname(dest_path))
         authorization = self.authorize_upload(project_name, bucket_name, dest_path)
         upload = Upload(file=file, file_size=size, upload_path=authorization.upload_path)
-        return self.upload_parts(upload, uuid.uuid4().hex, max_retries)
+        num_tries = max_retries
+        last_status_code = None
+        metis_uid = uuid.uuid4().hex
+        while num_tries > 0:
+            try:
+                return self.upload_parts(upload, metis_uid, last_status_code == 422)
+            except (RequestException, HTTPError) as e:
```

Comment:
Here's the problem, I think --- this `except` clause won't catch the `upload_parts` exceptions because this function is returning the generator, not running it.  You need the except inside `upload_parts`.

Reply:
In other words, when you call `upload_parts`, you don't run any part of upload parts at that point.
When upload_parts is running is during the `for _ in upload_parts()` or any `yield from` chained on it.

But, if your exception handling is inside upload_parts, then it DOES catch, since upload_parts itself is run in the generator context.

Reply:
try this on:

```
def m():
  print("hi!")
  yield 1

gen = m()
next(gen)
```